sanic_jwt/configuration.py

```diff
@@ -1,119 +1,313 @@
+        self._value = value
+
+    def __call__(self, **kwargs):
+        if asyncio.get_event_loop().is_running():
+            if self._get_from_config is not None:
+                args = []
+
+                if self._inject_request and is_cached("_request"):
+                    args.append(get_cached("_request"))
+                return self._get_from_config.__call__(*args)
+
+            if is_cached(self._item_name):
+                return get_cached(self._item_name)
+
+        return self._value
+
+    @property
+    def _get_from_config(self):
+        if hasattr(self._config, self._get_fn):
+            return getattr(self._config, self._get_fn)
+
+        return None
+
+    @property
+    def _get_fn(self):
+        return "get_{}".format(self._item_name)
+
+    @property
+    def aliases(self):
+        return self._aliases
 
 
 class Configuration:
+
+    def __iter__(self):  # noqa
+        for key in self.config_keys:
+            yield getattr(self, key)
+
+        for key in self.aliases_keys:
+            yield getattr(self, key)
+
+    def __contains__(self, item):
+        return item in self.all_config_keys
+
+    def __new__(cls, *args, **kwargs):
+        instance = super().__new__(cls)
+
+        _defaults = copy.deepcopy(defaults)
+        _aliases = copy.deepcopy(aliases)
+        _config_keys = []
+
+        if args and isinstance(args[0], dict):
+            _args = cls.extract_presets(args[0])
+            for key, value in _args.items():
+                if key in _defaults or key in _aliases:
+                    _defaults.update({key: value})
+
+        for key, value in _defaults.items():
+            item_aliases = []
+            if key in _aliases:
+                item_aliases = [aliases.get(key)]
+
+            # check if a configuration key is set and is an instance of ConfigItem
+            if (
+                hasattr(instance, key)
+                and isinstance(getattr(instance, key), ConfigItem)
+            ):
+                _update_config_item(key, item_aliases, instance)
+            # check if a configuration key is set with a value
+            elif hasattr(instance, key):
+                val = getattr(instance, key)
+                _create_or_overwrite_config_item(
+                    val, key, item_aliases, instance
+                )
+            else:
+                _create_or_overwrite_config_item(
+                    value, key, item_aliases, instance
+                )
+
+            # check if a setter is available on config class
+            fn_name = "set_{}".format(key)
+            if hasattr(instance, fn_name):
+                set_fn = getattr(instance, fn_name)
+                if not callable(set_fn):
+                    logger.warning(
+                        'variable "%s" set in Configuration is not callable',
+                        fn_name,
+                    )
+                    continue
+
+                val = set_fn.__call__()
```

Comment:
The unfortunate side effect of doing this in `__new__` is that it ends up being one of the few methods that we do not allow to be both a `callable` and `awaitable`.

The only way I could think to get around this would be with something like:

    await Configuration.setup(...)

Reply:
Not worth changing for now... just food for thought.

Reply:
Also ... now that we have `ConfigItem.update()` do we really need `set_fn` other than backwards compatibility? If it is not already configured this way... perhaps `set_fn` merely calls the `update` method (or the other way around).

Reply:
Yeap, I think this may be more straightforward.

> The unfortunate side effect of doing this in `__new__` is that it ends up being one of the few methods that we do not allow to be both a `callable` and `awaitable`.

Well, for most configurations I think a key is enough; some may require a callable for it, but I don't see any usage for any async method inside `Configuration`, since we don't get any instance running while `Configuration` is not fully instantiated ... Aside from async methods from a `Configuration` subclass ?